pkg/build/registry/buildconfig/etcd/etcd.go

```diff
@@ -47,3 +51,32 @@ func NewStorage(s storage.Interface) *REST {
 
 	return &REST{store}
 }
+
+func (r *REST) Delete(ctx kapi.Context, name string, options *kapi.DeleteOptions) (runtime.Object, error) {
+	bcObj, err := r.Get(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+	bc := bcObj.(*api.BuildConfig)
+
+	// Just update DeletionTimestamp so Builds get deleted
+	if bc.DeletionTimestamp.IsZero() {
+		now := unversioned.Now()
+		bc.DeletionTimestamp = &now
+		bcObj, _, err = r.Update(ctx, bc)
+		return bcObj, err
+	}
+
+	if !bc.Status.CanDelete {
+		err = kapierrors.NewConflict("BuildConfig", fmt.Sprintf("%s/%s", bc.Namespace, bc.Name), fmt.Errorf("The system is ensuring all Builds instantiated from this BuildConfig are removed. Upon completion, this BuildConfig will automatically be purged by the system."))
+		return nil, err
```

Comment:
well it may not have gotten deleted because yet because it's still in the process of being deleted.  you almost certainly don't want to retry it.

the point is, now that the buildconfig has a DeletionTimestamp, the buildconfig controller is going to attempt to delete it every time a resync interval hits.

let's go through the flow here:
1) buildconfig gets marked with a DeletionTimestamp
2) buildcontroller sees the buildconfig update
3) buildcontroller starts deleting child builds
4) if during this time, delete gets called again, it's not an error, nothing needs to happen at all.  the DeletionTimestamp is already set, but it's not an error to call delete on it again.
5) buildcontroller finishes deleting children, sets CanDelete on the BC and calls delete on the buildconfig
6) buildcontroller sees the update to CanDelete, ignores it.
7) REST api sees delete request, notes CanDelete is true, deletes it.

if something goes wrong in step 5, such that the BC never gets updated to CanDelete=true, or if the REST call to delete it fails, it'll get retried during the next list/sync loop in which the controller will process all the buildconfigs again and see a buildconfig that has a DeletionTimestamp, at which point it should pick up at step 2/3.

the problem scenario is when the buildcontroller tries to delete the buildconfig, but gets an error.  As you noted, that's going to result in a retry loop where HandleBuildConfig gets called again with the buildconfig object.  That's fine to a point, but that only retries for a minute or so and then gives up.  At that point you're going to be left with a buildconfig that has a DeletionTimestamp and has CanDelete=true, and it's never going to get deleted unless someone manually invokes the rest api to delete it again.